### PR TITLE
Added render class for Viewfield.

### PR DESCRIPTION
**Pull request by marcus-n3rd**
Absorbed the following functions into the class:
- viewfield_element_info()
- viewfield_pre_render()
- viewfield_post_render()
- _viewfield_get_view_args()
  -- This was made getViewArgs() which is a protected method

Left rendering for the theme, i.e.: removed `$output = render ($view_el);`, other than that I didn't touch any logic within the functions. 

Thank Metal Toad for paying my hours for this. Lightly tested this on a client's project and I'd say it's now eligible to go from sandbox to full project status.

### CODE REVIEW

**Comment by marcus-n3rd** [comment]
I combined the changes from:
- @seanbeaton 's https://github.com/derhasi/viewfield/pull/5
- @luksak 's https://github.com/derhasi/viewfield/pull/2

File: src/Render/Element/Viewfield.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\Core\Render\Element\Viewfield.
+ */
+
+namespace Drupal\Core\Render\Element;
```
Comment (derhasi) [inline src/Render/Element/Viewfield.php:8]: It should not use core's namespace, but `Drupal\viewfield\...` instead.

File: src/Render/Element/Viewfield.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+use Drupal\Core\Render\Element;
+
+/**
+ * Renders a view of the selected view.
+ *
+ * @RenderElement("viewfield")
+ */
+class Viewfield extends RenderElement {
+
+  public function getInfo() {
```
Comment (derhasi) [inline src/Render/Element/Viewfield.php:19]: This needs at least function documentation. I guess `{@inheritdoc}` should do its job here.

File: src/Render/Element/Viewfield.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        array($class, 'preRenderTextfield'),
+      ),
+      '#theme' => 'viewfield_formatter_default',
+      '#post_render' => array(
+        array($class, 'postRenderViewfield'),
+      ),
+    );
+  }
+
+  public static function preRenderViewfield($element) {
```
Comment (derhasi) [inline src/Render/Element/Viewfield.php:33]: This needs some proper function documentation and type declaration.

File: src/Render/Element/Viewfield.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      $current_path = $url->toString();
+      $element['#view']->override_path = $current_path;
+
+      // @todo Store views arguments serialized.
+      $element['#view_arguments'] = Viewfield::getViewArgs($element['#view_arguments'], $element['#entity_type'], $element['#entity']);
+    }
+    return $element;
+  }
+
+  public static function postRenderViewfield($content, $element) {
```
Comment (derhasi) [inline src/Render/Element/Viewfield.php:61]: This needs some proper function documentation and type declaration too.

File: src/Render/Element/Viewfield.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    $stack = &drupal_static('viewfield_stack', array());
+    unset($stack[$element['#entity_type']][$element['#entity_id']]);
+    return $content;
+  }
+
+  /**
+   * Perform argument replacement
+   */
+  protected function getViewArgs($vargs, $entity_type, $entity) {
```
Comment (derhasi) [inline src/Render/Element/Viewfield.php:70]: There needs to be some parameter documentation.

**Review by derhasi — CHANGES_REQUESTED** [review]
@marcus-n3rd, thanks for your work. I'm just starting getting into the issues again after some months.

Here are some issues I found from a "meta" perspective. I will have a closer look at the functionality itself later.

**Comment by marcus-n3rd** [comment]
Oh fudge. While step debugging through in order to double check the argument types, I found that none of Viewfield class is actually used. :facepalm: I'm going to run through and see if there are parts that can be consumed and used in the other classes.

**Comment by marcus-n3rd** [comment]
I am really pondering the value of having it as a RenderElement. Could the functionality just be moved into the ViewfieldDefaultFormatter (with the theme name changing to simply `viewfield` rather than `viewfield_formatter_default`)? Is the `viewfield_stack` static variable still needed with Drupal 8's render caching, because that would get rid of the need of the pre/post rendering functions/methods?

**Comment by luksak** [comment]
Well, I was trying to fix the caching issues and added that in the field formatter. As Berdir suggested, we should move that somewhere else where the render array is being built and that is what you approach does, right?

**Comment by marcus-n3rd** [comment]
Closing in preference of #8.

**Pull request closed without merging**